File: src/per/zzy/rayleigh/leetcode/java/Tree.java
package per.zzy.rayleigh.leetcode.java;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Tree {

    public static void main(String[] args) {

    }

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();

        if (root == null) return result;

        List<TreeNode> curLevel = new ArrayList<>();
        curLevel.add(root);

        while (!curLevel.isEmpty()) {
            List<TreeNode> temp = curLevel;
            curLevel = new ArrayList<TreeNode>();
            List<Integer> values = new ArrayList<>();
            for (TreeNode node : temp) {
                values.add(node.val);
                if (node.left != null) {
                    curLevel.add(node.left);
                }

                if (node.right != null) {
                    curLevel.add(node.right);
                }
            }

            result.add(values);
        }

        return result;
    }

    // 使用队列，优化
    public List<List<Integer>> levelOrder2(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();

        if (root == null) return result;

        Queue<TreeNode> curLevel = new LinkedList<>();
        curLevel.offer(root);

        while (!curLevel.isEmpty()) {
            List<Integer> values = new ArrayList<>();
            int size = curLevel.size();


            for (int i = 0; i < size; i++) {
                TreeNode node = curLevel.poll();
                if (node != null) {
                    values.add(node.val);

                    if (node.left != null) {
                        curLevel.offer(node.left);
                    }

                    if (node.right != null) {
                        curLevel.offer(node.right);
                    }

                }
            }
            result.add(values);
        }

        return result;
    }


}

File: src/per/zzy/rayleigh/leetcode/kotlin/Tree.kt
package per.zzy.rayleigh.leetcode.kotlin
object Tree {

    /**
     * Example:
     * var ti = TreeNode(5)
     * var v = ti.`val`
     * Definition for a binary tree node.
     * class TreeNode(var `val`: Int) {
     *     var left: TreeNode? = null
     *     var right: TreeNode? = null
     * }
     */

    fun levelOrder(root: TreeNode?): List<List<Int>> {
        val r = mutableListOf<List<Int>>()
        var curLevel = mutableListOf<TreeNode>()
        if (root == null) return r
        curLevel.add(root)

        while (curLevel.isNotEmpty()) {
            val temp = curLevel
            curLevel = mutableListOf<TreeNode>()

            val tempList = mutableListOf<Int>()
            for (node in temp) {
                tempList.add(node.v)
                node.left?.let {
                    curLevel.add(it)
                }
                node.right?.let {
                    curLevel.add(it)
                }
            }
            r.add(tempList)
        }

        return r
    }

}